src/server/scripts/Spells/spell_mage.cpp

```diff
@@ -695,6 +701,84 @@ class spell_mage_water_elemental_freeze : public SpellScript
     }
 };
 
+// Ice Lance - 30455
+class spell_mage_ice_lance : public SpellScript
+{
+    PrepareSpellScript(spell_mage_ice_lance);
+
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellInfo({
+            SPELL_MAGE_ICE_LANCE,
+            SPELL_MAGE_ICE_LANCE_TRIGGER
+        });
+    }
+
+    void HandleOnHit(SpellEffIndex /*effIndex*/)
+    {
+        Unit* caster = GetCaster();
+        Unit* target = GetHitUnit();
+        if (!caster && !target)
+            return;
+
+        bool targetFrozen = target->HasAuraState(AURA_STATE_FROZEN);
+
+        // Thermal Void
+        if (caster->HasAura(SPELL_MAGE_THERMAL_VOID) && (targetFrozen || caster->HasAura(SPELL_MAGE_FINGERS_OF_FROST)))
+        {
+            if (Aura* icyVeins = caster->GetAura(SPELL_MAGE_ICY_VEINS))
+            {
+                if (SpellInfo const* thermalVoidInfo = sSpellMgr->GetSpellInfo(SPELL_MAGE_THERMAL_VOID, GetCastDifficulty()))
+                {
+                    if (SpellEffectInfo const* eff0 = thermalVoidInfo->GetEffect(EFFECT_0))
+                    {
+                        int32 increaseDuration = eff0->CalcValue() * IN_MILLISECONDS;
+                        int32 newDuration = (icyVeins->GetDuration() + increaseDuration);
+                        icyVeins->SetDuration(newDuration);
+                    }
+                }
+            }
+        }
+
+        // if( target->HasAuraState(AURA_STATE_FROZEN))
+            caster->CastSpell(target, SPELL_MAGE_ICE_LANCE_TRIGGER, true); // TODO damage x3 on frozen target
+
+        // Splitting Ice
+        if (caster->HasAura(SPELL_MAGE_SPLITTING_ICE))
+        {
+            float radius = 10.0f;
+            std::list<Unit*> TargetList;
+            Trinity::AnyFriendlyUnitInObjectRangeCheck u_check(target, target, radius);
+            Trinity::UnitListSearcher<Trinity::AnyFriendlyUnitInObjectRangeCheck> searcher(target, TargetList, u_check);
+            Cell::VisitAllObjects(target, searcher, radius);
+
+            for (std::list<Unit*>::const_iterator itr = TargetList.begin(); itr != TargetList.end(); ++itr)
+            {
+                if ((*itr)->GetGUID() == target->GetGUID()) continue;
+                caster->CastSpell((*itr), SPELL_MAGE_ICE_LANCE_TRIGGER, true); // TODO 80% of the damage intead of 100%
+                break;
+            }
+        }
+
+        // Chain Reaction
+        if (targetFrozen || caster->HasAura(SPELL_MAGE_FINGERS_OF_FROST))
+        {
+            if (Aura* chainReaction_aura = caster->GetAura(SPELL_MAGE_CHAIN_REACTION))
```

Comment:
Don't search for existing aura to add a stack. Simply cast the spell again, it will automatically stack